text-transform:uppercase in SVG causes misplaced letters

Ahmad-S792 opened the pull request:
<pre>
text-transform:uppercase in SVG causes misplaced letters
<a href="https://bugs.webkit.org/show_bug.cgi?id=171664">https://bugs.webkit.org/show_bug.cgi?id=171664</a>
rdar://problem/31989978

Reviewed by NOBODY (OOPS!).

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: <a href="https://github.com/chromium/chromium/commit/829357862f40854d617b4b8640cbe526e6792db3">https://github.com/chromium/chromium/commit/829357862f40854d617b4b8640cbe526e6792db3</a>

The `originalText` function was added in 143476@main to account for any
text changes and then changed into map. This patch leverages it to apply
SVG whitespace rules as stated in web specification [1], which suggests to
look for CSS Whitespace handling [2] except when it is legacy cases (i.e.,
xml:space) [3]. Previously, we were not applying SVG whitespace rules to
account for `pre` whitespace as a result, when `text-transform` was used,
the transform text operations needs to follow [4] order, where `uppercase`
is first and it should happen after `white space` processing [5]
(Phase I: Collapsing and Transformation):

Additionally, it simplifies `styleDidChange` to ‘setText’ only
when `old` and `new` differ.

In future, we can simplify `applySVGWhitespaceRules` rules further as part
of bug 278524.

[1] <a href="https://svgwg.org/svg2-draft/text.html#TextWhiteSpace">https://svgwg.org/svg2-draft/text.html#TextWhiteSpace</a>
[2] <a href="https://drafts.csswg.org/css-text-3/#white-space-processing">https://drafts.csswg.org/css-text-3/#white-space-processing</a>
[3] <a href="https://svgwg.org/svg2-draft/text.html#LegacyXMLSpace">https://svgwg.org/svg2-draft/text.html#LegacyXMLSpace</a>
[4] <a href="https://drafts.csswg.org/css-text-3/#text-transform-order">https://drafts.csswg.org/css-text-3/#text-transform-order</a>
[5] <a href="https://drafts.csswg.org/css-text-3/#white-space-phase-1">https://drafts.csswg.org/css-text-3/#white-space-phase-1</a>

* Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
(WebCore::RenderSVGInlineText::originalText const):
(WebCore::RenderSVGInlineText::styleDidChange):
* LayoutTests/svg/text/text-transform-whitespace-rules.svg: Add Test Case
* LayoutTests/svg/text/text-transform-whitespace-rules-expected.svg: Add Test Case Expectation
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/748f1c8c5d72c6b1031517a6064d176a8152f6f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63839 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43196 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16436 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67860 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14447 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65959 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14727 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51439 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9989 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66908 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40008 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55265 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32056 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36688 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12652 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13320 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58658 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69556 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7786 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12520 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58762 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7819 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55363 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58908 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6475 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39016 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40095 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39838 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
```diff
@@ -89,7 +89,8 @@ String RenderSVGInlineText::originalText() const
 void RenderSVGInlineText::setRenderedText(const String& text)
 {
-    RenderText::setRenderedText(text);
+    const auto& textWS = applySVGWhitespaceRules(text, style().whiteSpace() == WhiteSpace::Pre);
```
The function returns a copy so this has to be `auto adjustedText =`

Ahmad-S792 replied:
@smfr - done in https://github.com/WebKit/WebKit/commit/65e9b8c9ee1294f4334e74a34bd52ba75e1d9dd5.

ojhunt replied:
@smfr with no warnings or anything?

Ahmad-S792 replied:
@smfr - I approached it via Blink's patch but I have another patch to fix another test case, which reduces scope of `applySVGWhitespaceRules` as well. So overall `makeString` operations in `applySVGWhitespaceRules` will be halved by another patch.

Ahmad-S792 replied:
@smfr - based on our discussion, I have updated commit message to reflect on why `whitespace` rules need to be applied before `text-transform` as per web-specification and also other bits.

mdubet commented on Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
```diff
@@ -89,7 +89,8 @@ String RenderSVGInlineText::originalText() const
 {
-    RenderText::setRenderedText(text);
+    auto adjustText = applySVGWhitespaceRules(text, style().whiteSpace() == WhiteSpace::Pre);
+    RenderText::setRenderedText(adjustText);
```
It's a bit sad that we can't move the string into `setRenderedText()`. We should add an overload.

Ahmad-S792 committed:
text-transform:uppercase in SVG causes misplaced letters

https://bugs.webkit.org/show_bug.cgi?id=171664
rdar://problem/31989978

Reviewed by NOBODY (OOPS!).

This patch aligns WebKit with Blink / Chromium and Gecko / Firefox.

Merge: https://github.com/chromium/chromium/commit/829357862f40854d617b4b8640cbe526e6792db3

The `originalText` function was added in 143476@main to account for any
text changes and then changed into map. This patch leverages it to apply
SVG whitespace rules as stated in web specification [1], which suggests to
look for CSS Whitespace handling [2] except when it is legacy cases (i.e.,
xml:space) [3].  Previously, we were not applying SVG whitespace rules to
account for `pre` whitespace as a result, when `text-transform` was used,
the transform text operations needs to follow [4] order, where `uppercase`
is first and it should happen after `white space` processing [5]
(Phase I: Collapsing and Transformation):

Additionally, it simplifies `styleDidChange` to ‘setText’ only
when `old` and `new` differ.

In future, we can simplify `applySVGWhitespaceRules` rules further as part
of bug 278524.

[1] https://svgwg.org/svg2-draft/text.html#TextWhiteSpace
[2] https://drafts.csswg.org/css-text-3/#white-space-processing
[3] https://svgwg.org/svg2-draft/text.html#LegacyXMLSpace
[4] https://drafts.csswg.org/css-text-3/#text-transform-order
[5] https://drafts.csswg.org/css-text-3/#white-space-phase-1

* Source/WebCore/rendering/svg/RenderSVGInlineText.cpp:
(WebCore::RenderSVGInlineText::originalText const):
(WebCore::RenderSVGInlineText::styleDidChange):
* LayoutTests/svg/text/text-transform-whitespace-rules.svg: Add Test Case
* LayoutTests/svg/text/text-transform-whitespace-rules-expected.svg: Add Test Case Expectation
